distributed/asyncio.py

```diff
@@ -0,0 +1,71 @@
+from __future__ import annotations
+
+import asyncio
+from collections.abc import AsyncIterator
+from contextlib import asynccontextmanager
+
+
+class RLock:
+    """Asyncio reentrant lock, which allows the same owner (or two owners that compare
+    as equal) to be inside the critical section at the same time.
+
+    Note that here the owner is an explicit, generic object, whereas in
+    :func:`threading.RLock` and :func:`multiprocessing.RLock` it's hardcoded
+    respectively to the thread ID and process ID .
+
+    **Usage**::
+
+        lock = RLock()
+        async with lock("my-owner"):
+            ...
+
+    **Tip**
+
+    You can mix reentrant and non-reentrant owners; all you need to do is create an
+    owner that doesn't compare as equal to other instances of itself::
+
+        lock = RLock()
+
+        async def non_reentrant():
+            async with lock(object()):
```

Comment:
the `object()` sentinel feels a bit odd. Why not just call it without an argument for this purpose?

Reply:
I see how you are using it below. if the object is within a tuple, that makes more sense for this use case.

If we make the module private, I don't care about this